### PR TITLE
Lesson 02: Implementing a character device

### PR DESCRIPTION
**Pull request by MaksymPrymierov**
* Implement character device driver, and control interfaces for this device
* Add script to test this device `test.sh`
* Add testing instruction `Testing.md`

### CODE REVIEW

File: 02-KernelModules_CharacterDevice/Makefile [hunk 1]
```diff
@@ -0,0 +1,18 @@
...
+else
+
+.PHONY: all clean
+all:
+	${MAKE} -C ${BUILD_KERNEL} M=${CURDIR} modules
+
+clean:
+	${MAKE} -C ${BUILD_KERNEL} M=${CURDIR} clean
+
+endif
```
Comment (galkinletter) [inline 02-KernelModules_CharacterDevice/Makefile:18]:
```suggestion
endif

```

Please add empty line at the end of file

File: 02-KernelModules_CharacterDevice/concdev.h [hunk 2]
```diff
@@ -0,0 +1,16 @@
...
+void remove_cdevice(void);
+
+int cdev_open(struct inode *, struct file *);
+int cdev_release(struct inode *, struct file *);
+void clear_buffer(void);
+
+ssize_t cdev_read(struct file *filp, char __user *buf, size_t count, loff_t *f_pos);
+ssize_t cdev_write(struct file *filp, const char __user *buf, size_t count, loff_t *f_pos);
+
+#endif
```
Comment (galkinletter) [inline 02-KernelModules_CharacterDevice/concdev.h:16]: Plaese add empty line at the end of file

File: 02-KernelModules_CharacterDevice/concmod.c [hunk 3]
```diff
@@ -0,0 +1,51 @@
...
+
+	return 0;
+
+	err:
+		cmod_exit();
+		return status;
+}
+
+module_init(cmod_init);
+module_exit(cmod_exit);
```
Comment (galkinletter) [inline 02-KernelModules_CharacterDevice/concmod.c]: add empty line at the end of file

File: 02-KernelModules_CharacterDevice/proc_info.c [hunk 4]
```diff
@@ -0,0 +1,71 @@
...
+		retval = -EFAULT;
+		goto err;
+	}
+	*pos += count;
+
+	return count;
+
+	err:
+		return retval;
+}
```
Comment (galkinletter) [inline 02-KernelModules_CharacterDevice/proc_info.c]: add empty line at the end of file

File: 02-KernelModules_CharacterDevice/proc_info.h [hunk 5]
```diff
@@ -0,0 +1,17 @@
...
+	size_t *capacity;
+	size_t *buffer_size;
+	const char *name;
+};
+
+int create_proc_info(const char *name, size_t *buffer_size, size_t *capacity);
+void remove_proc_info(void);
+static ssize_t proc_info_read(struct file *filp, char *buf, size_t count, loff_t *pos);
+
+#endif
```
Comment (galkinletter) [inline 02-KernelModules_CharacterDevice/proc_info.h]: Please add empty line at the end of file

File: 02-KernelModules_CharacterDevice/sysfs_control.c [hunk 6]
```diff
@@ -0,0 +1,78 @@
...
+	if (buffer) {
+		clear_function();
+		printk("MESSAGE (SysFsControl): buffer has been cleared\n");
+	} else {
+		printk("MESSAGE (SysFsControl): buffer cannot be cleared\n");
+
+	}
+
+	return count;
+}
```
Comment (galkinletter) [inline 02-KernelModules_CharacterDevice/sysfs_control.c]: Please add empty line at the end of file

File: 02-KernelModules_CharacterDevice/sysfs_control.h [hunk 7]
```diff
@@ -0,0 +1,10 @@
+#ifndef SYSFS_CONTROL_H
+#define SYSFS_CONTROL_H
+
+int create_sysfs_control(const char *name, void (*clear) (void), dev_t devt);
+void remove_sysfs_control(void);
+static ssize_t sysfs_control_store(struct device *dev,
+								   struct device_attribute *attr, 
+								   const char *buf, size_t count);
+
+#endif
```
Comment (galkinletter) [inline 02-KernelModules_CharacterDevice/sysfs_control.h]: add empty line at the end of file

File: 02-KernelModules_CharacterDevice/test.sh [hunk 8]
```diff
@@ -0,0 +1,69 @@
...
+	cat /dev/concdev
+elif [[ $1 == "din" ]]; then
+	echo $2 >> /dev/concdev
+elif [[ $1 == "memout" ]]; then
+	memout
+elif [[ $1 == "clear" ]]; then
+	echo 1 >> /sys/devices/concdev/clear
+else
+	help
+fi
```
Comment (galkinletter) [inline 02-KernelModules_CharacterDevice/test.sh]: Please add empty line at the end of file

File: 02-KernelModules_CharacterDevice/sysfs_control.c [hunk 9]
```diff
@@ -0,0 +1,78 @@
...
+	device_destroy(char_dev_class, devt_buf);
+	class_destroy(char_dev_class);
+
+	sysfs_remove_file(&ko, &attr_chard.attr);
+	root_device_unregister(char_device);
+
+	printk("MESSAME (SysFsControl): sysfs control removed\n");
+}
+
+static ssize_t sysfs_control_store(struct device *dev,
```
Comment (galkinletter) [inline 02-KernelModules_CharacterDevice/sysfs_control.c]:
```suggestion
static ssize_t sysfs_control_store(struct device *dev,struct device_attribute *attr, const char *buf, size_t count)
```

Need fixe code style

**Review by galkinletter — CHANGES_REQUESTED** [review]
Need fixe code style

File: 02-KernelModules_CharacterDevice/Makefile (hunk 1 above)
Comment (galkinletter) [reply 02-KernelModules_CharacterDevice/Makefile:18]: add empty line at the end of file

**Review by galkinletter — CHANGES_REQUESTED** [review]
Need fix empty line at the end of file

**Comment by connor41** [comment]
скрин

**Comment by connor41** [comment]
i.imgur.com/nafE5kt.png

**Comment by MaksymPrymierov** [comment]
image

File: 02-KernelModules_CharacterDevice/concdev.c [hunk 10]
```diff
@@ -0,0 +1,181 @@
...
+
+	result = create_sysfs_control(DEVICE_NAME, clear_buffer, dev);
+
+	if (result < 0) {
+		goto fail;
+	}
+
+	return 0;
+	
+	fail:
```
Comment (oleg-h2o) [inline 02-KernelModules_CharacterDevice/concdev.c:90]: Error handling is not just printing a message.
remove_proc_info() ?
unregister_chrdev_region(devno, count_dev) ?
cdev_del(cdev) ?
etc.

**Review by oleg-h2o — CHANGES_REQUESTED** [review]
Fix error handling.

File: 02-KernelModules_CharacterDevice/concdev.c [hunk 11]
```diff
@@ -0,0 +1,182 @@
...
+	.open = cdev_open,
+	.release = cdev_release,
+	.read = cdev_read,
+	.write = cdev_write,
+};
+
+int create_cdevice(int major, int minor, int count, size_t size)
+{
+	int result;
+    dev_t dev = 0;
```
Comment (galkinletter) [inline 02-KernelModules_CharacterDevice/concdev.c]:
```suggestion
         dev_t dev = 0;
```
Please fix last problem in  code style

**Review by galkinletter — CHANGES_REQUESTED** [review]
Need fix last problem in  code style

**Comment by galkinletter** [comment]
Accepted!

**Pull request merged**